### ISSUE TITLE
Request: display stats in a historical format

### ISSUE THREAD

**Issue by phemmer** [moby/moby#10770]
Currently `docker stats` uses a top-like (curses) interface, but I think it would be far more useful to have historical output.

For example:

```
# docker stats deadbeef012 5 3
          CPU%   Mem%  Mem(M)  Limit(M)  NetIn(K)  NetOut(K)
12:47:04  32.4    5.4    54.3    100.0        387     32,344
12:47:09  36.2    5.9    59.2    100.0        455     75,923
12:47:14  21.2    4.2    41.9    100.0        147     12,036
```

This output is very easy to scan through (the current stats output is rather cluttered and hard to read), and shows you how the container is trending.
If you want to monitor multiple containers, it would simply add another column after the time for container ID.
This is very similar to the behavior of existing statistics utilities such as `sar` and `iostat`. This also makes the output able to be parsed by scripts.

The command line format is `docker stats CID [INTERVAL [COUNT]]`, where `INTERVAL` is the period over which to gather stats, and `COUNT` is how many intervals to gather before exiting.

**Comment by brahmaroutu** [comment]
@crosbymichael 
When I tried this few observation I had and hence I am holding off on a PR. Firstly for implementing a interval, subscribers receive events every 1 sec which is hardcoded and I do not know how to change that. I am fixing the issue from the client side.
Secondly, if we want to enforce number of lines, I believe we should do so only if we are querying a single container stats, the output can become unreadable IMO. I may be wrong.
sudo docker stats --lines=10  --interval=5 a09639dd03a4 (works fine)
sudo docker stats --lines=10  --interval=5 a09639dd03a4 37f934a5947c (it works like is today, ignoring lines=10)

**Comment by cpuguy83** [comment]
I think that makes sense only if you are running it in the background and you later want to read the history of stats like you would with sar, etc.

Why not implement this separately using the API?

**Comment by brahmaroutu** [comment]
@cpuguy83 Thanks, For this cli command, I think this fix should work well. A new Rest endpoint which can control poll frequency (I was suggesting in issue #10711) would be good, I can add a secondary API endpoint if it makes sense and a  new cli command. But I am not sure how much new functionality it is bringing in, Like formatting(ability to identify and group), event definitions, etc. If not It may look same as this API.

**Comment by cpuguy83** [comment]
@brahmaroutu I'm referring to implementing outside of Docker with the PR.

**Comment by crosbymichael** [comment]
The worry is over loading the cli trying to cater to many different metric and stats requirements.  The API should be easily consumable and give you much more flexibility and even more metrics than what is able to be displayed via the CLI.  

Is there a reason why consuming the API directly is a problem in your use case?

**Comment by cpuguy83** [comment]
This would be pretty easy to implement at this point since the API can be polled for a single stat now.
Do we want to implement this in the CLI or leave to other tools?

**Comment by phemmer** [comment]
The only thing I might have against "other tools" is that this command would be meant for human consumption. Most other tools for docker that I've seen are meant for automation or to perform a task. E.G. advanced configuration, health monitoring, integration with other services, etc. I can't recall any utilities whose primary purpose are for a human to interact with docker directly.

And then I think it becomes awkward to have to do `docker foo` for stuff that's built-in, and then `docker-bar` for the custom commands, especially if it's something used frequently. Maybe if docker had the ability to pick up custom commands and provide them as `docker bar` (ala git style), it might not be so bad.

**Comment by icecrime** [comment]
I think this should be solved by #25820.

**Comment by thaJeztah** [comment]
I think we can close this one, now that there's a `--no-stream` option, as well as options to format as JSON (`--format json`); combinations thereof could be used to collect historical data.

```bash
docker container stats --no-stream
CONTAINER ID   NAME      CPU %     MEM USAGE / LIMIT     MEM %     NET I/O       BLOCK I/O         PIDS
d2b8b34790f3   foo       0.00%     9.766MiB / 7.662GiB   0.12%     1.04kB / 0B   7.77MB / 8.19kB   11
```

```bash
docker container stats --format json --no-stream
{"BlockIO":"7.77MB / 12.3kB","CPUPerc":"0.00%","Container":"d2b8b34790f3","ID":"d2b8b34790f3","MemPerc":"0.12%","MemUsage":"9.746MiB / 7.662GiB","Name":"foo","NetIO":"1.18kB / 0B","PIDs":"11"}

docker container stats --no-trunc --format json --no-stream
{"BlockIO":"7.77MB / 12.3kB","CPUPerc":"0.00%","Container":"d2b8b34790f3","ID":"d2b8b34790f3b0730452b36386120779089835fd4f5f68e2e830b673aa1ff99e","MemPerc":"0.12%","MemUsage":"9.746MiB / 7.662GiB","Name":"foo","NetIO":"1.32kB / 0B","PIDs":"11"}
```
